### PR TITLE
`hs project dev` add polling to project fetch in case of 403

### PR DESCRIPTION
**Pull request by adamawang**
## Description and Context

This PR specifically deals with fetchProject behavior when creating a new sandbox with `hs project dev`. Due to gating slowness (sometimes) when creating a new sandbox in prod, we are receiving 403's on initial fetch with a forbidden request error. 

We are adding a pollingFetchProject util that keeps retrying (up to max 30s) until a 404 or a success is received. This also ensures that the portal is ungated if the user needs to create a new project in the next step. 

We can remove this polling piece once CRM dev tools are ungated to all. 

## Screenshots

Adding a spinner to show progress
Screenshot 2023-05-09 at 15 20 23

Once we get a 404 or success, we remove the spinner
Screenshot 2023-05-09 at 15 20 39

### CODE REVIEW

File: packages/cli/lib/projects.js
Comment (adamawang) [inline packages/cli/lib/projects.js:180]: Try first for all other usages of `ensureProjectExists`, only poll if response is not successful or a 404

File: packages/cli/lib/projects.js
Comment (adamawang) [inline packages/cli/lib/projects.js:191]: Rather than succeed, I thought removing would be cleaner since we already have a warning message: 
Screenshot 2023-05-09 at 15 20 39

Let me know what you think

File: packages/cli/lib/projects.js
Comment (brandenrodgers) [inline packages/cli/lib/projects.js:171]: Going off of the screenshot here, it was my understanding that the issue was that we are not able to successfully create the project in the newly-created sandboxes. This seems to be adding polling to the fetching of the project, not the creation of it in the target account. Let me know if I'm just misunderstanding what the original issue was!

Screenshot 2023-05-08 at 15 53 30

Comment (adamawang) [reply packages/cli/lib/projects.js:171]: Yep its both! The very last prompt was asking me to create a project, but then it fails and exits. This is before you merged in your other branch with an error message though. 

By adding the polling to the fetch, we can ensure that upon a 404 or success, the portal has been successfully ungated and project create in the next step will work

File: packages/cli/lib/projects.js
Comment (brandenrodgers) [inline packages/cli/lib/projects.js:197]: Should we also abort on 500 status codes? Just to cover our bases

File: packages/cli/lib/projects.js
Comment (brandenrodgers) [inline packages/cli/lib/projects.js:215]: I'm thinking that we should only use this new polling logic for the Dev Mode work. Maybe we can add a new `withPolling` option to this arguments in `ensureProjectExists`. I don't think this issue will have much of an effect on the existing project-related commands.

Comment (adamawang) [reply packages/cli/lib/projects.js:215]: Yep exactly - I made it so behavior for all other usages stayed the same, but I like the prop idea better

**Comment by brandenrodgers** [comment]
I tested this locally and it's all working correctly 👌 I'll take a quick look, but this should be good to go after you get the updates in.

File: packages/cli/lib/projects.js
Comment (adamawang) [reply packages/cli/lib/projects.js:171]: Yeah if its not ungated, we'll get a 403 with the following message: 
Screenshot 2023-05-08 at 15 28 17

**Comment by adamawang** [comment]
Updated in my last commit - tested with `hs project dev`, `hs project watch` and `hs project upload` to make sure `ensureProjectExists` works as intended, looking good on my end

**Review by brandenrodgers — APPROVED** [review]
🚢

**Pull request merged**
